start github-and-away tech note (#1167)
ropensci/roweb3: content/blog/2025-11-07-github-and-away/index.md

---

> title: Code hosting options beyond GitHub

Reviewer: [Style Guide](https://blogguide.ropensci.org/authortechnical.html#styleguide) and (Ahem!) the author's [checklist](https://blogguide.ropensci.org/authorchecklistany.html)
```suggestion
title: Code Hosting Options Beyond GitHub
```

author: 

> - Mark Padgham

Reviewer: @noamross needs to be here somewhere, author or editor? If you added content, I'd say author.

editor:
date: '2025-11-07'
slug: beyond-github
description: How to manage and mirror code repositories across different platforms.
output: hugodown::md_document
tags:
  - tech notes
params:
  doi: "10.59350/bhr5x-d3429"
editor:
- Maëlle Salmon
- Adam Sparks
- Eric Scott

> - Doug Kelkhoff

Reviewer: Doug needs an [author file](https://blogguide.ropensci.org/authortechnical.html#createauthorfile)

---

rOpenSci makes heavy use of GitHub for our projects and services, including [software peer-review](https://github.com/ropensci/software-review/issues?q=sort%3Aupdated-desc%20is%3Aissue%20state%3Aclosed).
GitHub is by far the most widely used git or code-hosting platform, and the combination of its popularity and freemium services have made it central to open-source and R communities.
However, for a variety of reasons, some of our community members or potential members may prefer or need to use other platforms.

> These reasons may include concerns about privacy, including [identifying information required for age verification in some jurisdictions](https://www.abc.net.au/news/2025-09-24/digital-dilemna-social-media-age-ban-platforms/105807302). Other reasons include a desire to support platforms with different ownership or business models, based in other countries, or supprting alternatives to avoid the risk of hegemony.

Reviewer: I like this section, it's absolutely needed and helpful especially for people like me that might have a hazy idea but would like more.
```suggestion
These reasons may include concerns about privacy, including [identifying information required for age verification in some jurisdictions](https://www.abc.net.au/news/2025-09-24/digital-dilemna-social-media-age-ban-platforms/105807302). Other reasons include a desire to support platforms with different ownership or business models, based in other countries, or supporting alternatives to avoid the risk of hegemony.
```

## Mirroring on other platforms

Different remote instances of a Git repository are often referred to as "mirrors". Here "mirror" refers to copies of code on any platforms other than the primary remote.
Most examples presume GitHub to be the primary remote, because rOpenSci operations are built around that presumption, but any platform can serve as a primary remote location, with GitHub, and any number of other platforms then being mirrors.
"Alternative locations" just implies any additional mirror locations other than the primary remote.

Code mirrors are only ever updated through `git push` events from local versions, or from other `git` events on the primary remote.
Change events originating on the primary remote are generally incorporated in a local version via `git pull`, and then pushed out to all mirrored versions.
A `git pull` command should only ever be applied to the primary remote version, and never to any alternative mirror versions.
In this diagram, the large yellow arrow represents the only connection where both `push` and `pull` events are allowed.
All other arrows are `push` events only.

> ![](./local-remote.png)

Author: Oh yeah, sorry, that's just so that I can see them on my local deploy which doesn't render the moustachiod-ones. All duplicated ones will be removed prior to publication.

{{< figure src = "local-remote.png" alt = "Git interaction directions between local and remote repositories." class = "pull-left" caption = "Git interaction directions between local and remote repositories.">}}

> ### Mirroring on Codeberg or GitLab

Reviewer: ```suggestion
## Migrating to Codeberg or GitLab
```

The easiest platforms on which to mirror your code are currently [Codeberg](https://codeberg.org) and [GitLab](https://gitlab.com), both of which have inbuilt options to mirror repositories directly from a large number of other platforms, including GitHub.
Both of these platforms can also mirror additional information including issues, issue labels, pull requests, and releases.
There are some limitations. For instance, GitHub's auto-links within issues and pull requests may be lost.

To mirror a repo on Codeberg, click the large "+" button on the top right of the main menu bar, and select "New migration", like this:

> ![](codeberg-new-migration.png)

Reviewer: ```suggestion
To migrate a repo to Codeberg, click the large "+" button on the top right of the main menu bar, and select "New migration".
```

{{< figure src = "codeberg-new-migration.png" alt = "Codeberg new migration button." class = "pull-left" caption = "Codeberg new repository migration button.">}}

That will then open up the following grid of options from where you want to mirror your repository:

![](codeberg-migration-options.png)

{{< figure src = "codeberg-migration-options.png" alt = "Codeberg repository migration options." class = "pull-left" caption = "Codeberg repository migration options.">}}

> To migrate from GitHub, click the symbol to open a migration to fill in some details, where you can also paste a GitHub token into "access token", and mirror almost all other aspects, including issues, pull requests, and releases.

Reviewer: ```suggestion
To migrate from GitHub, click the symbol to open a migration to fill in some details, where you can also paste a GitHub token into "access token" (ensuring that it has at least read access to all aspects you want transferred), and migrate almost all other aspects, including issues, pull requests, and releases.
```

Note that the migration process may take 10 minutes or more.

On GitLab, the equivalent "+" button is on the top left, where "_New project/repository_" should be clicked:

> {{< figure src = "gitlab-new.png" alt = "GitLab new repository button." class = "pull-left" caption = "GitLab new repository button.">}}

Reviewer: ```suggestion
On GitLab, the equivalent "+" button is on the top left, where "_New project/repository_" should be clicked.
```

> {{< figure src = "gitlab-new.png" alt = "GitLab new repository button." class = "pull-left" caption = "GitLab new repository button.">}}

Reviewer: ```suggestion
{{< figure src = "gitlab-new.png" alt = "" class = "pull-left" caption = "GitLab new repository button.">}}
```

Both Codeberg and GitLab enable all information on all issues and pull requests to be transferred across, although generally without hyperlinks other than those those linking within or between issues of the same repository.
All other hyperlinks, including any to GitHub users, or to other repositories, will be lost.
The transfer will be of information at the time the "New migration" on Codeberg or "New import" on GitLab is initiated.
From that time point on, any activity via issues or pull requests on any hosting platform will be specific to that platform only.[^2]

[^2]: There are tools for continuous mirroring such as [GitHubCodebergMirror](https://codeberg.org/wl/GithubCodebergMirror), but there is likely some fragility in these setups.

To transfer issues from GitHub to Codeberg, first click on the "GitHub" symbol in the "Migrate repository" options shown above, after which appears a dialog with options for what you would like migrated.
These options include "issues", "Pull requests", "Labels", "Milestones", and "Releases," also with the clear information that,

> > Access token is required to migrate additional items

Reviewer: ```suggestion
No other platforms currently offer the one-click migrate functionality of Codeberg or GitLab.
To migrate in all other cases, you'll need to:
```

For that, you just need to paste a GitHub token in the Codeberg field, ensuring that the token has at least read access to all aspects you want transferred.
GitLab transfers issues, pull requests, tags, and releases by default, with even more possible through provision of a GitHub access token.
In both cases, transfer can take some time (10-20 minutes or more), following which all issues will appear exactly as on GitHub, but generally minus the hyperlinks described above.

> Other code hosting platforms like [SourceHut](https://sr.ht/), or the distributed hosting system [tangled](https://tangled.org/) do not provide facilities for directly transferring or mirroring issues or pull requests (and in fact generally don't even have such things, as they really are specific to the GitHub platform; [SourceHut](https://sr.ht), for example, implements its own systems of "todos" and "patches".)

Reviewer: Move some of this to the section on migration to Codeberg and GitLab (see those comments above) and to the section on Mirroring elsewhere (see the comments in that section).

```suggestion

### A note on transferring issues

Issues, pull requests, and other information on GitHub can be transferred both to Codeberg and GitLab, but only when a repository is first migrated.

Both Codeberg and GitLab enable all information on all issues and pull requests to be transferred across, although generally without hyperlinks other than those those linking within or between issues of the same repository.
All other hyperlinks, including any to GitHub users, or to other repositories, will be lost.
The transfer will be of information at the time the "New migration" on Codeberg or "New import" on GitLab is initiated.
From that time point on, any activity via issues or pull requests on any hosting platform will be specific to that platform only.[^2]

[^2]: There are tools for continuous mirroring such as [GitHubCodebergMirror](https://codeberg.org/wl/GithubCodebergMirror), but there is likely some fragility in these setups.
```

## Managing one repository across multiple platforms

To be clear, this entire post is about moving away from having code hosted on a single platform, to distributing across multiple platforms through _mirroring_.
If, for example, you only want to migrate away from GitHub to some single, other platform, then both Codeberg and GitLab already offer the full solutions described above.

> We nevertheless recommend mirroring across multiple platforms, to reduce risks associated with dependence on any single platform.

Reviewer: What are these risks?

> We nevertheless recommend mirroring across multiple platforms, to reduce risks associated with dependence on any single platform.

Reviewer: ```suggestion
The goal here is to move from having code hosted on a single platform to distributing code across multiple platforms through _mirroring_. This reduces the risks associated with dependence on any single platform, and also allows you to use an alternative platform to GitHub while still benefiting with GitHub operations used by other organizations (like rOpenSci!).
Therefore the next step is to setup mirroring which involves two parts, 
adding remotes and then synchronizing those remotes.
```

All other remote origins should be considered `push` mirrors only, and never `pull`.
In the rare case that conflicts from other sources arise, you may need to `git push --force` to _other_ remotes (or the [safer version of `git push --force-with-lease`](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force-with-leaserefnameexpect)).
You should never `git push --force` to the main branch of your primary source.

### Adding remotes to your local Git information

For each additional remote source, you'll need to add a remote URL with [`git remote add`](https://git-scm.com/docs/git-remote).
There are many ways to do this.
The pure Git way of managing multiple remote sources is to take advantage of `git remote set-url --add` to add additional URLs to a single remote identifier.
[This blog post](https://jeffkreeftmeijer.com/git-multiple-remotes/) details how to do that safely, to ensure only one primary remote is configured to `fetch`, while allowing `push` events to all others.
An alternative option would be to initially create an additional remote like `git remote add other https://codeberg.org/ropensci/my-package`.

> You can then extend that with each additional remote URL with `set-url --add`.

Reviewer: I don't understand this line, how does this work?

Running `git push other <branch>` will then push that branch to all remote URLs specified in `other`.

> Yet another approach is to define a [custom command](https://stackoverflow.com/questions/60060217/how-do-i-make-custom-git-commands) for `git push` to call a local script.

Reviewer: Is this section really necessary? I find it complex and a bit confusing and possibly only useful to people trying to mass migrate & mirror a bunch of repositories. Could it be put in an appendix at the end perhaps?

Author: Thanks, I've made it a footnote now.

This can be done by adding some "gitbin" directory to a default system `PATH` variable (in `.bashrc`, for example, as `export PATH="$PATH:$HOME/.gitbin"`).
Git will automatically recognize any scripts within `$PATH` named with the prefix "git-".
It is best to name locally-defined Git commands differently to standard Git commands.
For example, a file named `git-pushall` placed in that folder will be called by the `git pushall` command.
The file itself could contain any of the options described in [this StackOverflow answer](https://stackoverflow.com/a/18674313).
Even more arcane alternatives include my own [git push bash script](https://github.com/mpadge/dotfiles/blob/main/system/gitpush.bash) which recreates the now obsolete push-by-password functionality of GitHub, while also pushing to all other listed remote sources.

### Synchronizing multiple remotes

> As with `git remote` setups described above, there are several ways to synchronize multiple remote instances of a single repository.

Reviewer: ```suggestion
As with the `git remote` setups described above, there are several ways to synchronize multiple remote instances of a single repository.
```

Author: Couldn't commit here, but done nonetheless. Thanks

The pure Git way is to use local commands like those described and linked to above, to `git push` changes to all remote locations.
(This is the approach recommended by GitHub, but GitHub is an anomaly among code hosting platforms in offering no easy ability to interact with other platforms.)

> Both Codeberg and GitLab offer mirroring services to enable these sites to be used as a primary site, and to mirror all code changes directly across to any alternative site such as GitHub.

Reviewer: ```suggestion
However, both Codeberg and GitLab offer mirroring services to enable these sites to be used as a primary site, and to mirror all code changes directly across to alternative sites such as GitHub.
```

If your primary home is _not_ GitHub, then you can create an additional GitHub-specific `.github/README.md` file which will then be displayed on GitHub only, while all other platforms will display the root README contents.
An example is Doug Kelkhoff's [`github.com/dgkf/options` package](https://github.com/dgkf/options) which has a [`.github/README.md`](https://github.com/dgkf/options/tree/main/.github) explaining that GitHub hosts a read-only mirror of the repository, and directing people to the primary location at [`codeberg.org/dgkf/options`](https://codeberg.org/dgkf/options).
That primary location then displays the full contents of the root README file.
This trick only works for GitHub - all other platforms display the root README file at all times.

## Words of warning

Whatever approach you decide to explore, you're bound to realise that managing code distributed across multiple platforms is more difficult that on a single site.

> Many features you might be used on GitHub, for example, may not be implemented on other platforms, or may have other names and inconsistent behaviours.

Reviewer: ```suggestion
Whatever approach you decide to explore, you're bound to realise that managing code distributed across multiple platforms is more difficult than on a single site.
Many features you might be used to on GitHub, for example, may not be implemented on other platforms, or may have other names and inconsistent behaviours.
```

  To mirror releases on GitLab, you need to.
    - `git push --tags gitlab <branch_name>`, just like on Codeberg, to explicitly push all Git tags.
    - Manually go through each tag and issue a new release, for which you can retrospectively set the release data to be the same historical date as the original GitHub release date.

There are many other ways small and large by which different code-hosting platforms and systems differ from one another.
Ongoing maintenance across different platforms will always present challenges, but we hope to have given you enough information here to get you started.

## rOpenSci repos on Codeberg, GitLab, or elsewhere

If you're an author of an rOpenSci package, you can follow the steps described above to mirror your code to any location on Codeberg, GitLab, or to any other platform.
Please let us know if you'd like any of these platforms other than GitHub to act as the primary home for your code, and we'll update our internal data accordingly.
We recommend moving repositories either to personal namespaces on these platforms, or to other organizational namespaces to which you have administrative access.

> We also maintain the organizations [Codeberg/ropensci](https://codeberg.org/ropensci) or [GitLab/ropensci](https://gitlab.com/ropensci).

Reviewer: ```suggestion
We also maintain the organizations [Codeberg/ropensci](https://codeberg.org/ropensci) and [GitLab/ropensci](https://gitlab.com/ropensci).
```
